File: jetbrains/JetBrains.RenderDoc.RdClient/RenderDocClient.cs
using JetBrains.Annotations;
using JetBrains.Collections.Viewable;
using JetBrains.Lifetimes;
using JetBrains.Rd;
using JetBrains.Rd.Impl;
using JetBrains.RenderDoc.RdClient.Model;

namespace JetBrains.RenderDoc.RdClient;

[PublicAPI]
public class RenderDocClient
{
    public async Task Start(Lifetime lifetime, long sessionId)
    {
        var protocolId = $"RenderDocClient::{sessionId}";

        var scheduler = DefaultScheduler.Instance;
        var taskScheduler = scheduler.AsTaskScheduler();
        var host = new RenderDocHost();
        var port = await host.Start(lifetime);
        
        var serializers = new Serializers(taskScheduler, null);
        RenderDocModel.RegisterDeclaredTypesSerializers(serializers);
        
        var client = new SocketWire.Client(lifetime, scheduler, port, protocolId);
        var protocol = new Protocol(protocolId, serializers, new Identities(IdKind.Client), scheduler, client, lifetime);
        await lifetime.Start(taskScheduler, () =>
        {
            var model = new RenderDocModel(lifetime, protocol);
            model.Pong.Advise(lifetime, _ =>
            {
                Console.WriteLine("Pong!!!");
            });
            model.Ping();
            model.Ping();
            model.Shutdown();
        });
        await host.HostExitCode;
    }
}

File: jetbrains/src/main/kotlin/com/jetbrains/renderdoc/rdClient/RenderDocClient.kt
package com.jetbrains.renderdoc.rdClient

import com.jetbrains.rd.framework.*
import com.jetbrains.rd.util.PublicApi
import com.jetbrains.rd.util.lifetime.Lifetime
import com.jetbrains.rd.util.reactive.IPropertyView
import com.jetbrains.rd.util.reactive.IScheduler
import com.jetbrains.renderdoc.rdClient.model.RenderDocModel
import com.jetbrains.renderdoc.rdClient.model.renderDocModel

@PublicApi
class RenderDocClient(lifetime: Lifetime, @PublicApi val scheduler: IScheduler, sessionId: Long, port: Int) {
    companion object {
        suspend fun createWithHost(lifetime: Lifetime, scheduler: IScheduler, sessionId: Long, binDir: String): RenderDocClient {
            val hostLifetimeDef = lifetime.createNested()
            val host = RenderDocHost(hostLifetimeDef, binDir)

            val clientLifetimeDef = hostLifetimeDef.createNested()
            return RenderDocClient(clientLifetimeDef, scheduler, sessionId, host.getPort())
        }
    }

    @PublicApi
    val connected: IPropertyView<Boolean>
    @PublicApi
    val model: RenderDocModel

    init {
        val protocolId = "RenderDocClient::$sessionId"
        val client = SocketWire.Client(lifetime, scheduler, port, protocolId)
        val protocol = Protocol(protocolId, Serializers(), Identities(IdKind.Client), scheduler, client, lifetime)
        RenderDocModel.register(protocol.serializers)

        connected = client.connected
        model = protocol.renderDocModel
    }
}